Supporting multilingual data

UladBohdan opened the pull request:
* `Article` model splitted into `Article` (common data) and `LocalizedArticle` (localized data)

A suggested workflow for Frontend:
* creating new `Article`: `/api/articles/` - returns `_id`
* adding a localization: `/api/articles/localize/:id` - returns localized data. `locale` is in the body and is a part of `LocalizedArticle` model.
* fixing a localization: PUT `/api/articles/localize/:localizationSlug`

Drapegnik commented on src/api/article/article.model.js:
```diff
@@ -5,22 +5,16 @@ import omit from 'lodash/omit';
     ref: 'User',
   },
-  tags: [String],
+  locales: [
```
Array? not mapping `locale` -> `ArticleData`?

UladBohdan replied:
I chose array to have `ref: 'ArticleData',` explicitly specified in the model.
Otherwise it's Schema.Types.Mixed and it's not clear at all if it's a map or whatever else.
Frontend should be fine with processing arrays: each `ArticleData` object has a `locale` field (yes, it makes it `O(N)`).

Drapegnik commented on src/api/article/controller.js:
```diff
@@ -42,23 +44,33 @@ export const getAll = ({ query, user }, res, next) => {
+  // TODO(uladbohdan): to fix for unexisting article.
+  let articleData;
+  try {
+    articleData = await ArticleData.findOne({ slug }).exec();
```
I think with promise chain this method would look more clean

UladBohdan replied:
Same as below: I don't know how to implement a chain if we work with different models at the same time.

Drapegnik replied:
```js

Drapegnik commented on src/api/article/data/controller.js:
```diff
@@ -0,0 +1,31 @@
+export const create = async ({ params: { articleId }, body }, res, next) => {
+  let article;
+  try {
+    article = await Article.findOne({ _id: articleId }).exec();
```
please, use promises here

UladBohdan replied:
you mean instead of try-catch? I have no idea how to implement that. Can you help?

Drapegnik commented on src/api/article/data/index.js:
```diff
@@ -0,0 +1,14 @@
+const router = Router();
+
+router.post('/:articleId', requireAuth, verifyPermission('canCreateArticle'), controller.create);
+router.put('/:slug', requireAuth, verifyPermission('canCreateArticle'), controller.update);
```
What about using `router.put('/:articleId/:slug'`? Thus, it is more clearly that create and put are applied to different entities. (usually `post on /` - create, `put on /:id` - edit)

Drapegnik replied:
Lol, how it can work at all? `controller.update` will never be called!

UladBohdan replied:
It will be if http query is `PUT` rather than `POST`.

`/:articleId/:slug` seems to have redundant information. We'll need additional checks if the `slug` is really a part of the `articleId`.

UladBohdan commented:
@Drapegnik thanks for you help with chaining! Not sure why but had troubles with implementation

UladBohdan commented:
@babajka/developers the PR approaches the merge! Please comment, it's a lot of model refactoring.

Drapegnik commented on src/api/article/article.model.js:
```diff
@@ -58,8 +48,9 @@ const ArticleSchema = new Schema({
-  ...omit(article.toObject(), ['_id', '__v']),
+  ...omit(article.toObject(), ['__v']),
   brand: article.brand.name,
+  locales: keyBy(article.locales, 'locale'),
```
👍

UladBohdan replied:
lodash is awesome

Drapegnik commented on src/api/article/controller.js:
```diff
@@ -58,7 +65,7 @@ export const create = async ({ body }, res, next) => {
     await articleBrand.save();
 
-    const articleCollection = await ArticleCollection.findOne({ slug: body.collectionSlug });
+    const articleCollection = await ArticleCollection.findOne({ slug: body.collectionSlug }).exec();
```
Doesn't this work without `exec`? What is difference?

UladBohdan replied:
It works without `exec`. The difference is without exec it's a Mongoose Query (not a Promise) and after we call an exec() it's a real Promise. What do we need to await it then?

Drapegnik replied:
ok, lets use `exec`

Drapegnik commented on src/api/article/controller.js:
```diff
@@ -87,13 +94,25 @@ export const create = async ({ body }, res, next) => {
+  LocalizedArticle.findOne({ slug })
+    .then(checkIsFound)
+    .then(({ articleId }) =>
+      Article.findOneAndUpdate({ _id: articleId }, { active: false }, { new: true })
```
I think you can use just `update`

Drapegnik commented on src/api/article/controller.js:
```diff
@@ -87,13 +94,25 @@ export const create = async ({ body }, res, next) => {
+    .then(({ articleId }) =>
+      Article.findOneAndUpdate({ _id: articleId }, { active: false }, { new: true })
+    )
     .then(checkIsFound)
```
it's redundant

Drapegnik commented on src/api/article/localized/controller.js:
```diff
@@ -0,0 +1,31 @@
+
+import LocalizedArticle from './model';
+
+export const create = async ({ params: { articleId }, body }, res, next) =>
```
Redundant `async` cause you don't use `await` inside. Please check other methods too.

Drapegnik commented on src/api/article/localized/controller.js:
```diff
@@ -0,0 +1,31 @@
+import LocalizedArticle from './model';
+
+export const create = async ({ params: { articleId }, body }, res, next) =>
+  Article.findOne({ _id: articleId })
```
```

Drapegnik commented on src/api/article/localized/controller.js:
```diff
@@ -0,0 +1,31 @@
+        .then(sendJson(res))
+        .catch(next);
+    })
+    .catch(next);
```
```js
async create(...) => {
	try {
		const article = await Article.findOne({ _id: articleId });
		checkIsFound(article);
	} catch err {
}
```

UladBohdan replied:
it's probably more readable but I still not sure when and why do we want to rewrite such structure into a chain from try-catch or back.
I think it's fine as it's written already.

Drapegnik commented on src/api/article/localized/controller.js:
```diff
@@ -0,0 +1,31 @@
+          return data;
+        })
+        .then(data => {
+          article.save();
```
Are you sure that article saved before that response is send?

UladBohdan replied:
`save()` returns a Promise. Should I **await** it to be sure it's saved before it's sent?

Drapegnik replied:
yep

Drapegnik commented on src/api/article/localized/tests.js:
```diff
@@ -0,0 +1,95 @@
+    promises.push(Article.remove());
+    promises.push(LocalizedArticle.remove());
+    promises.push(User.remove());
+    await Promise.all(promises);
```
What about:

Drapegnik commented on src/api/article/localized/controller.js:
```diff
@@ -1,13 +1,26 @@
+    .populate('locales', 'locale')
     .then(checkIsFound)
+    .then(article => {
+      article.locales.forEach(loc => {
```
```js
if (article.locales.includes(body.locale)) {
	throw new HttpError(...)
}
```

UladBohdan replied:
it's not the same: `article.locales` here is an array of object, each containing a `locale` field. I can't *include* over it.

Drapegnik replied:
```js
article.locales.map(({ locale }) => locale).includes(body.locale)

UladBohdan commented:
I've also removed try-catch structures in `auth` and `users` tests - they seem to be completely useless. Correct me if they had their role.

@Drapegnik I guess I've fixed it all

Drapegnik commented on src/api/article/index.js:
```diff
@@ -4,21 +4,28 @@ import { requireAuth, verifyPermission } from 'auth';
 
 import brandRoutes from 'api/article/brand';
 import collectionRoutes from 'api/article/collection';
+import dataRoutes from 'api/article/localized';
```
old name `dataRoutes`

Drapegnik commented on src/utils/testing.js:
```diff
@@ -0,0 +1,5 @@
+
+import mongoose from 'mongoose';
+
+export const dropData = () => mongoose.connection.db.dropDatabase();
```
:+1:
